src/envs/pybullet_blocks.py

```diff
@@ -0,0 +1,754 @@
+        relatively or absolutely. The rel_or_abs argument indicates
+        whether each dimension is relative ("rel") or absolute ("abs").
+        """
+
+        def _get_current_and_target_pose(
+                state: State, objects: Sequence[Object],
+                params: Array) -> Tuple[Pose3D, Pose3D]:
+            robot, = objects
+            # De-normalize parameters to actual table coordinates.
+            x_norm, y_norm = params
+            x = self.x_lb + (self.x_ub - self.x_lb) * x_norm
+            y = self.y_lb + (self.y_ub - self.y_lb) * y_norm
+            current_pose = (state[robot][0], state[robot][1], state[robot][2])
+            table_pose = (x, y, self.table_height)
+            target_pose = self._convert_rel_abs_to_abs(table_pose,
+                                                       rel_or_abs_target_pose,
+                                                       rel_or_abs)
+            return current_pose, target_pose
+
+        return self._move_end_effector_to_pose(name, types, params_space,
+                                               _get_current_and_target_pose)
+
+    def _change_fingers(self, name: str, target_val: float,
+                        types: Sequence[Type],
+                        params_space: Box) -> ParameterizedOption:
+        """Creates a ParameterizedOption for changing the robot fingers."""
+
+        assert types[0] == self._robot_type
+
+        def _policy(state: State, memory: Dict, objects: Sequence[Object],
+                    params: Array) -> Action:
+            del memory, objects, params  # unused
+            current_val = state.get(self._robot, "fingers")
+            finger_action = target_val - current_val
+            finger_action = np.clip(finger_action, self.action_space.low[3],
+                                    self.action_space.high[3])
+            return Action(
+                np.array([0., 0., 0., finger_action], dtype=np.float32))
+
+        def _terminal(state: State, memory: Dict, objects: Sequence[Object],
+                      params: Array) -> bool:
+            del memory, objects, params  # unused
+            current_val = state.get(self._robot, "fingers")
+            squared_dist = (target_val - current_val)**2
+            return squared_dist < self._grasp_tol
+
+        return ParameterizedOption(name,
+                                   types=types,
+                                   params_space=params_space,
+                                   policy=_policy,
+                                   initiable=lambda _1, _2, _3, _4: True,
+                                   terminal=_terminal)
+
+    def _get_end_effector_move_toward_waypoint_action(
+            self, ee_pose: Pose3D, target_pose: Pose3D) -> Action:
+        action = self._move_gain * np.subtract(target_pose, ee_pose)
+        action_norm = np.linalg.norm(action)  # type: ignore
+        if action_norm > self._max_vel_norm:
```

Comment:
so suppose that `action` is `[100, 1, 0]`. if we removed this if statement, then the return would be `[0.05, 0.05, 0]`. but instead, we would want it to be closer to `[0.049, 0.01, 0]` or something. (right?) because we want to move in the right direction, toward the target pose